File: ships-preview/src/main/terminal/security.ts
/**
 * Terminal Security Module
 * 
 * Command validation and sanitization for safe terminal execution.
 * This is the security gate - all commands must pass through here.
 */

import { existsSync, statSync } from 'fs';
import type { CommandValidationResult, AllowedCommand } from './types';

/**
 * Whitelisted command prefixes with their configurations.
 * Only these commands can be executed.
 */
export const ALLOWED_COMMANDS: AllowedCommand[] = [
  { prefix: 'npm', description: 'Node package manager', maxTimeout: 300000, requiresApproval: false },
  { prefix: 'npx', description: 'NPM package runner', maxTimeout: 300000, requiresApproval: false },
  { prefix: 'yarn', description: 'Yarn package manager', maxTimeout: 300000, requiresApproval: false },
  { prefix: 'pnpm', description: 'PNPM package manager', maxTimeout: 300000, requiresApproval: false },
  { prefix: 'git', description: 'Git version control', maxTimeout: 60000, requiresApproval: false },
  { prefix: 'python', description: 'Python interpreter', maxTimeout: 60000, requiresApproval: true },
  { prefix: 'py', description: 'Python interpreter (Windows)', maxTimeout: 60000, requiresApproval: true },
  { prefix: 'pip', description: 'Python package manager', maxTimeout: 300000, requiresApproval: true },
  { prefix: 'node', description: 'Node.js runtime', maxTimeout: 60000, requiresApproval: true },
];

/**
 * Blocked patterns that indicate dangerous operations.
 * Commands containing these will be rejected.
 */
const BLOCKED_PATTERNS: RegExp[] = [
  /rm\s+-rf/i,           // Recursive delete
  /rm\s+--force/i,       // Force delete
  /del\s+\/[fqs]/i,      // Windows force delete
  /rmdir\s+\/s/i,        // Windows recursive delete
  /format\s+/i,          // Disk format
  /mkfs\s+/i,            // Make filesystem
  /\|\s*sh/i,            // Pipe to shell
  /\|\s*bash/i,          // Pipe to bash
  /\|\s*cmd/i,           // Pipe to cmd
  /;\s*rm/i,             // Command chain to rm
  /&&\s*rm/i,            // Command chain to rm
  /sudo\s+/i,            // Privilege escalation
  /chmod\s+/i,           // Permission changes
  /chown\s+/i,           // Ownership changes
  /curl.*\|\s*sh/i,      // Curl pipe to shell
  /wget.*\|\s*sh/i,      // Wget pipe to shell
  /eval\s+/i,            // Eval execution
  /exec\s+/i,            // Exec execution
  /`.*`/,                // Backtick execution
  /\$\(/,                // Subshell execution
  />.*\/etc\//i,         // Write to /etc
  />.*\.bashrc/i,        // Write to bashrc
  />.*\.profile/i,       // Write to profile
];

/**
 * Blocked subcommands for specific tools.
 */
const BLOCKED_SUBCOMMANDS: Record<string, string[]> = {
  npm: ['exec', 'x'],  // npm exec can run arbitrary code
  git: ['filter-branch', 'gc', 'prune'],  // Destructive git commands
};

/**
 * Validate that a path is safe for command execution.
 */
export function isPathSafe(projectPath: string): boolean {
  if (!projectPath || !existsSync(projectPath)) {
    return false;
  }

  try {
    const stats = statSync(projectPath);
    if (!stats.isDirectory()) {
      return false;
    }

    // Block system directories
    const normalizedPath = projectPath.toLowerCase().replace(/\\/g, '/');
    const forbidden = [
      'c:/windows',
      'c:/program files',
      'c:/programdata',
      '/system',
      '/usr/bin',
      '/bin',
      '/etc',
      '/var',
    ];

    return !forbidden.some(f => normalizedPath.startsWith(f));
  } catch {
    return false;
  }
}

/**
 * Get the allowed command config for a command prefix.
 */
export function getAllowedCommandConfig(command: string): AllowedCommand | undefined {
  const cmdBase = command.trim().split(/\s+/)[0].toLowerCase();
  return ALLOWED_COMMANDS.find(ac => ac.prefix === cmdBase);
}

/**
 * Validate a command for safe execution.
 * This is the main security gate.
 */
export function validateCommand(command: string, cwd: string): CommandValidationResult {
  // Trim and normalize
  const trimmedCommand = command.trim();
  
  if (!trimmedCommand) {
    return { isValid: false, error: 'Empty command' };
  }

  // Check CWD is safe
  if (!isPathSafe(cwd)) {
    return { isValid: false, error: 'Invalid or unsafe working directory' };
  }

  // Get command base
  const parts = trimmedCommand.split(/\s+/);
  const cmdBase = parts[0].toLowerCase();
  const subCommand = parts[1]?.toLowerCase();

  // Check if command prefix is allowed
  const allowedConfig = ALLOWED_COMMANDS.find(ac => ac.prefix === cmdBase);
  if (!allowedConfig) {
    return { 
      isValid: false, 
      error: `Command '${cmdBase}' is not in the whitelist. Allowed: ${ALLOWED_COMMANDS.map(c => c.prefix).join(', ')}` 
    };
  }

  // Check for blocked subcommands
  const blockedSubs = BLOCKED_SUBCOMMANDS[cmdBase];
  if (blockedSubs && subCommand && blockedSubs.includes(subCommand)) {
    return { isValid: false, error: `Subcommand '${cmdBase} ${subCommand}' is blocked for security` };
  }

  // Check for blocked patterns
  for (const pattern of BLOCKED_PATTERNS) {
    if (pattern.test(trimmedCommand)) {
      return { isValid: false, error: 'Command contains blocked pattern' };
    }
  }

  // Check for path traversal attempts
  if (trimmedCommand.includes('..')) {
    return { isValid: false, error: 'Path traversal (..) is not allowed' };
  }

  // Check for environment variable manipulation in npm scripts
  if (cmdBase === 'npm' && subCommand === 'run') {
    // npm run is allowed, but we trust the project's package.json
    // This is a calculated risk - user controls their own project
  }

  return { 
    isValid: true, 
    sanitizedCommand: trimmedCommand 
  };
}

File: ships-backend/app/terminal/security.py
"""
Terminal Security Module

Command validation and whitelisting for safe terminal execution.
This is the security gate - all commands must pass validation.
"""

import re
import os
from pathlib import Path
from typing import Tuple, Optional, List
import logging

from .models import AllowedCommand

logger = logging.getLogger("ships.terminal")


# ============================================================================
# WHITELISTED COMMANDS
# ============================================================================

ALLOWED_COMMANDS: List[AllowedCommand] = [
    AllowedCommand(prefix="npm", description="Node package manager", max_timeout=300, requires_approval=False),
    AllowedCommand(prefix="npx", description="NPM package runner", max_timeout=300, requires_approval=False),
    AllowedCommand(prefix="yarn", description="Yarn package manager", max_timeout=300, requires_approval=False),
    AllowedCommand(prefix="pnpm", description="PNPM package manager", max_timeout=300, requires_approval=False),
    AllowedCommand(prefix="git", description="Git version control", max_timeout=60, requires_approval=False),
    AllowedCommand(prefix="python", description="Python interpreter", max_timeout=60, requires_approval=True),
    AllowedCommand(prefix="py", description="Python interpreter (Windows)", max_timeout=60, requires_approval=True),
    AllowedCommand(prefix="pip", description="Python package manager", max_timeout=300, requires_approval=True),
    AllowedCommand(prefix="node", description="Node.js runtime", max_timeout=60, requires_approval=True),
]


# ============================================================================
# BLOCKED PATTERNS
# ============================================================================

BLOCKED_PATTERNS: List[re.Pattern] = [
    re.compile(r"rm\s+-rf", re.IGNORECASE),           # Recursive delete
    re.compile(r"rm\s+--force", re.IGNORECASE),       # Force delete
    re.compile(r"del\s+/[fqs]", re.IGNORECASE),       # Windows force delete
    re.compile(r"rmdir\s+/s", re.IGNORECASE),         # Windows recursive delete
    re.compile(r"format\s+", re.IGNORECASE),          # Disk format
    re.compile(r"mkfs\s+", re.IGNORECASE),            # Make filesystem
    re.compile(r"\|\s*sh", re.IGNORECASE),            # Pipe to shell
    re.compile(r"\|\s*bash", re.IGNORECASE),          # Pipe to bash
    re.compile(r"\|\s*cmd", re.IGNORECASE),           # Pipe to cmd
    re.compile(r";\s*rm", re.IGNORECASE),             # Command chain to rm
    re.compile(r"&&\s*rm", re.IGNORECASE),            # Command chain to rm
    re.compile(r"sudo\s+", re.IGNORECASE),            # Privilege escalation
    re.compile(r"chmod\s+", re.IGNORECASE),           # Permission changes
    re.compile(r"chown\s+", re.IGNORECASE),           # Ownership changes
    re.compile(r"curl.*\|\s*sh", re.IGNORECASE),      # Curl pipe to shell
    re.compile(r"wget.*\|\s*sh", re.IGNORECASE),      # Wget pipe to shell
    re.compile(r"eval\s+", re.IGNORECASE),            # Eval execution
    re.compile(r"`.*`"),                              # Backtick execution
    re.compile(r"\$\("),                              # Subshell execution
    re.compile(r">\s*/etc/", re.IGNORECASE),          # Write to /etc
    re.compile(r">\s*~/.bashrc", re.IGNORECASE),      # Write to bashrc
    re.compile(r">\s*~/.profile", re.IGNORECASE),     # Write to profile
]

BLOCKED_SUBCOMMANDS: dict = {
    "npm": ["exec", "x"],  # npm exec can run arbitrary code
    "git": ["filter-branch", "gc", "prune"],  # Destructive git commands
}


# ============================================================================
# FORBIDDEN PATHS
# ============================================================================

FORBIDDEN_PATHS = [
    "c:\\windows",
    "c:\\program files",
    "c:\\programdata",
    "/system",
    "/usr/bin",
    "/bin",
    "/etc",
    "/var",
]


# ============================================================================
# VALIDATION FUNCTIONS
# ============================================================================

def is_path_safe(project_path: str) -> Tuple[bool, str]:
    """
    Validate that a path is safe for command execution.
    
    Returns:
        Tuple of (is_safe, error_message)
    """
    if not project_path:
        return False, "No project path provided"
    
    resolved = Path(project_path).resolve()
    
    if not resolved.exists():
        return False, f"Path does not exist: {project_path}"
    
    if not resolved.is_dir():
        return False, f"Path is not a directory: {project_path}"
    
    # Check forbidden paths
    normalized = str(resolved).lower().replace("\\", "/")
    for forbidden in FORBIDDEN_PATHS:
        if normalized.startswith(forbidden.replace("\\", "/")):
            return False, f"Cannot execute commands in system directory: {forbidden}"
    
    # Check it's not the backend directory
    backend_dir = Path(__file__).resolve().parent.parent.parent
    if str(resolved).startswith(str(backend_dir)):
        return False, "Cannot execute commands in the backend directory"
    
    return True, ""


def get_allowed_command_config(command: str) -> Optional[AllowedCommand]:
    """Get the config for an allowed command prefix."""
    cmd_base = command.strip().split()[0].lower()
    for ac in ALLOWED_COMMANDS:
        if ac.prefix == cmd_base:
            return ac
    return None


def validate_command(command: str, cwd: str) -> Tuple[bool, str, Optional[str]]:
    """
    Validate a command for safe execution.
    
    Returns:
        Tuple of (is_valid, error_message, sanitized_command)
    """
    trimmed = command.strip()
    
    if not trimmed:
        return False, "Empty command", None
    
    # Validate working directory
    path_safe, path_error = is_path_safe(cwd)
    if not path_safe:
        return False, path_error, None
    
    # Parse command
    parts = trimmed.split()
    cmd_base = parts[0].lower()
    sub_command = parts[1].lower() if len(parts) > 1 else None
    
    # Check whitelist
    allowed_config = get_allowed_command_config(trimmed)
    if not allowed_config:
        allowed_list = ", ".join(ac.prefix for ac in ALLOWED_COMMANDS)
        return False, f"Command '{cmd_base}' is not allowed. Whitelist: {allowed_list}", None
    
    # Check blocked subcommands
    blocked_subs = BLOCKED_SUBCOMMANDS.get(cmd_base, [])
    if sub_command and sub_command in blocked_subs:
        return False, f"Subcommand '{cmd_base} {sub_command}' is blocked for security", None
    
    # Check blocked patterns
    for pattern in BLOCKED_PATTERNS:
        if pattern.search(trimmed):
            return False, "Command contains blocked pattern", None
    
    # Check path traversal
    if ".." in trimmed:
        return False, "Path traversal (..) is not allowed", None
    
    logger.info(f"[SECURITY] ✅ Command validated: {trimmed[:50]}...")
    return True, "", trimmed
